File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        //base case 
        //if matrix empty
        if (matrix.length == 0 || matrix[0].length == 0) {
            return false;
        }
        int n = matrix.length;
        int m= matrix[0].length;
        int lo=0;
        int hi= (n*m)-1;

while(lo<=hi){
    int mid=(lo + (hi-lo)/2);
    if(matrix[mid/m][mid%m]==target){
        return true;
    }
    //search in right half
    if(matrix[mid/m][mid%m]<target){
        lo=mid+1;
    }else{
        //search in left half
        hi= mid-1;
    }
}return false;
    }
}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    const n = matrix.length;
    const m = matrix[0].length;
    let low=0;
    let high = n*m-1;
    while(low<=high){
        const mid = Math.floor((low+high)/2);
        const row = Math.floor(mid / m);
        const col = mid % m;

        if (matrix[row][col] === target) {
            return true;
        } else if (matrix[row][col] < target) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }

    return false;

};